models/demos/falcon7b/tt/falcon_attention.py

```diff
@@ -288,58 +339,90 @@ def forward(
+                    )
+                )
                 query_layer[i].deallocate()
                 key_layer_transposed[i].deallocate()
 
-        for i in range(self.num_devices):
-            attn_weights[i] = tt_lib.tensor.bcast(
-                attn_weights[i],
-                self.scalar[i],
-                tt_lib.tensor.BcastOpMath.MUL,
-                tt_lib.tensor.BcastOpDim.HW,
-                output_mem_config=self.model_config["PRE_SOFTMAX_SCALE_OUTPUT_MEMCFG"],
-            )
-
-        if attention_mask is not None:
+        if llm_mode == "prefill" or self.model_config["l1_sharded"] == False:
             for i in range(self.num_devices):
-                attn_weights[i] = tt_lib.tensor.add(
+                attn_weights[i] = tt_lib.tensor.bcast(
                     attn_weights[i],
-                    attention_mask[i],
-                    output_mem_config=self.model_config["PRE_SOFTMAX_MASK_OUTPUT_MEMCFG"],
+                    self.scalar[i],
+                    tt_lib.tensor.BcastOpMath.MUL,
+                    tt_lib.tensor.BcastOpDim.HW,
+                    output_mem_config=self.model_config["PRE_SOFTMAX_SCALE_OUTPUT_MEMCFG"],
                 )
 
-        ###############
-        ### SOFTMAX ###
-        ###############
-        # TODO: Replace with scaled_softmax_attention_mask from BERT
-        for i in range(self.num_devices):
-            attn_weights[i] = tt_lib.operations.primary.softmax_in_place(
-                attn_weights[i],
-            )
+            if attention_mask is not None:
+                for i in range(self.num_devices):
+                    attn_weights[i] = tt_lib.tensor.add(
+                        attn_weights[i],
+                        attention_mask[i],
+                        output_mem_config=self.model_config["PRE_SOFTMAX_MASK_OUTPUT_MEMCFG"],
+                    )
+            ###############
+            ### SOFTMAX ###
+            ###############
+            for i in range(self.num_devices):
+                attn_weights[i] = tt_lib.operations.primary.transformers.scale_mask_softmax_in_place(attn_weights[i])
+        else:
+            ###############
+            ### SOFTMAX ###
+            ###############
+            for i in range(self.num_devices):
+                attn_weights[i] = tt_lib.operations.primary.transformers.scale_mask_softmax_in_place(
+                    attn_weights[i],
+                    scale=self.scale,
+                    mask=attention_mask[i],
+                    program_config=tt_lib.operations.primary.transformers.SoftmaxShardedMultiCoreProgramConfig(
+                        compute_with_storage_grid_size=(8, 4),
+                        subblock_w=1,
+                        block_h=self.padded_local_heads // 32,
+                        block_w=padded_layer_past_len // 32,
+                    ),
+                    is_causal_mask=True,
```

Comment:
I don't think is_causal_mask should be enabled for decode

Reply:
is_causal_mask does not make a difference if the attention_mask is the same size as the input